File: FastPay/lib/assistant/local-rag/template.ts
import type { RetrievedLocalChunk } from "../types";

function dedupeActions(
  actions: { label: string; href: string }[],
): { label: string; href: string }[] {
  const seen = new Set<string>();
  return actions.filter((action) => {
    if (seen.has(action.href)) {
      return false;
    }
    seen.add(action.href);
    return true;
  });
}

export function chunksToTemplateReply(chunks: RetrievedLocalChunk[]): {
  reply: string;
  sources: { title: string; source: string; route?: string }[];
  actions: { label: string; href: string }[];
} {
  const top = chunks.slice(0, 3);

  if (!top.length) {
    return {
      reply:
        "I could not find FastPay docs for that. Try asking about transfers, bills, savings, loans, or KYC.",
      sources: [],
      actions: [],
    };
  }

  const lines = top.map(
    (chunk) => `• ${chunk.title ?? chunk.source}: ${chunk.text.slice(0, 220)}…`,
  );

  return {
    reply: ["Here is what I found in FastPay:", ...lines, "", "This is product guidance, not personal financial advice."].join(
      "\n",
    ),
    sources: top.map((chunk) => ({
      title: chunk.title ?? chunk.source,
      source: chunk.source,
      route: chunk.route,
    })),
    actions: dedupeActions(
      top
        .filter((chunk) => chunk.route || chunk.actionRoute)
        .slice(0, 2)
        .map((chunk) => ({
          label: `Open ${chunk.title ?? "screen"}`,
          href: chunk.actionRoute ?? chunk.route!,
        })),
    ),
  };
}

File: fastpay-assistant-ml/fastpay_assistant/rag/template.py
from __future__ import annotations

from fastpay_assistant.ml.action_ranker import dedupe_actions
from fastpay_assistant.types import ChatAction, ChatSource, RetrievedLocalChunk


def chunks_to_template_reply(
    chunks: list[RetrievedLocalChunk],
) -> tuple[str, list[ChatSource], list[ChatAction]]:
    top = chunks[:3]
    if not top:
        return (
            "I could not find FastPay docs for that. Try asking about transfers, bills, savings, loans, or KYC.",
            [],
            [],
        )

    lines = [f"• {c.title or c.source}: {c.text[:220]}…" for c in top]
    reply = "\n".join(
        [
            "Here is what I found in FastPay:",
            *lines,
            "",
            "This is product guidance, not personal financial advice.",
        ]
    )
    sources = [
        ChatSource(title=c.title or c.source, source=c.source, route=c.route) for c in top
    ]
    actions = dedupe_actions(
        [
            ChatAction(
                label=f"Open {c.title or 'screen'}",
                href=c.action_route or c.route or "/support",
            )
            for c in top
            if c.route or c.action_route
        ][:2]
    )
    return reply, sources, actions
